src/lib/coda_base/pending_coinbase.ml

```diff
@@ -0,0 +1,499 @@
+            in
+            let%bind h1 = Pedersen.Checked.Digest.choose_preimage h1
+            and h2 = Pedersen.Checked.Digest.choose_preimage h2 in
+            Pedersen.Checked.digest_triples
+              ~init:Hash_prefix.coinbase_merkle_tree.(height)
+              (to_triples h1 @ to_triples h2)
+
+          let assert_equal h1 h2 = Field.Checked.Assert.equal h1 h2
+
+          let if_ = Field.Checked.if_
+        end)
+        (struct
+          include Stack
+
+          type value = t [@@deriving sexp]
+
+          let hash = Checked.digest
+        end)
+
+    let depth = coinbase_tree_depth
+
+    type path = Pedersen.Digest.t list
+
+    module Address = struct
+      include Merkle_tree.Address
+
+      let typ = typ ~depth
+    end
+
+    type _ Request.t +=
+      | Coinbase_stack_path : Address.value -> path Request.t
+      | Get_coinbase_stack : Address.value -> (Stack.t * path) Request.t
+      | Set_coinbase_stack : Address.value * Stack.t -> unit Request.t
+      | Find_index_of_newest_stack : bool -> Address.value Request.t
+      | Find_index_of_oldest_stack : Address.value Request.t
+
+    let reraise_merkle_requests (With {request; respond}) =
+      match request with
+      | Merkle_tree.Get_path addr ->
+          respond (Delegate (Coinbase_stack_path addr))
+      | Merkle_tree.Set (addr, stack) ->
+          respond (Delegate (Set_coinbase_stack (addr, stack)))
+      | Merkle_tree.Get_element addr ->
+          respond (Delegate (Get_coinbase_stack addr))
+      | _ -> unhandled
+
+    let get t addr =
+      handle
+        (Merkle_tree.get_req ~depth (Hash.var_to_hash_packed t) addr)
+        reraise_merkle_requests
+
+    let%snarkydef update_stack t ~is_new_stack stack_before stack_after =
+      let%bind addr =
+        request_witness Address.typ
+          As_prover.(
+            map (read Boolean.typ is_new_stack) ~f:(fun s ->
+                Find_index_of_newest_stack s ))
+      in
+      handle
+        (Merkle_tree.modify_req ~depth (Hash.var_to_hash_packed t) addr
+           ~f:(fun stack ->
+             let%map () =
+               Coinbase_stack.Stack.assert_equal stack_before stack
+             in
+             stack_after ))
+        reraise_merkle_requests
+      >>| Hash.var_of_hash_packed
+
+    let%snarkydef delete_stack t stack_before stack_after =
+      let%bind addr =
+        request_witness Address.typ
+          As_prover.(map (return ()) ~f:(fun _ -> Find_index_of_oldest_stack))
+      in
+      handle
+        (Merkle_tree.modify_req ~depth (Hash.var_to_hash_packed t) addr
+           ~f:(fun stack ->
+             let%map () =
+               Coinbase_stack.Stack.assert_equal stack_before stack
+             in
+             stack_after ))
+        reraise_merkle_requests
+      >>| Hash.var_of_hash_packed
+  end
+
+  type t = {tree: Merkle_tree.t; pos_list: Stack_id.t list; new_pos: Stack_id.t}
+  [@@deriving sexp, bin_io]
+
+  let create_exn () =
```

Comment:
what does this function do?

Reply:
Creates a merkle tree with empty stacks. This function should go probably in `Sparse_ledger`